src/main/java/org/testng/providers/IObjectProvider.java

```diff
@@ -0,0 +1,29 @@
+package org.testng.providers;
+
+import java.lang.annotation.Annotation;
+import java.util.Map;
+import org.testng.IClass;
+import org.testng.ITestObjectFactory;
+import org.testng.internal.annotations.IAnnotationFinder;
+import org.testng.xml.XmlTest;
+
+public interface IObjectProvider {
+
+  Object create(Class<?> declaringClass,
+      Map<Class<?>, IClass> classes,
+      XmlTest xmlTest,
+      IAnnotationFinder finder,
+      ITestObjectFactory factory,
+      boolean create,
+      String errorMsgPrefix);
+
+  Object create(IClass iClass, Class<?> clazz);
+
+  default Object create(Class<?> clazz) {
+    return null;
+  }
+
+  boolean supportsDependencyInjection();
```

Comment:
As said before, the usage of the object provider should be the same with or without dependency injection.

Reply:
Yes, agreed but only if the object provider decides on the logic to be used internally. In our case, it is TestNG that takes a call on how to go about instance creation. TestNG is merely relying on the providers to provide different ways of producing objects. At the end, it is TestNG that would decide on the best way to create objects and hence this method.

Reply:
> At the end, it is TestNG that would decide on the best way to create objects and hence this method.

Sorry but I disagree. The purpose of the object creation separations should be for internal needs first, and for 3rd party integrations later.
Currently, we have 2 ways for object creation and the core should not care about it because it trusts the creation part.
The SPI will come later when the design will be stabilized, right?